tracer/src/Datadog.Trace/Util/UnmanagedMemoryPoolFactory.cs

```diff
@@ -0,0 +1,64 @@
+// <copyright file="UnmanagedMemoryPoolFactory.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Threading;
+using Datadog.Trace.AppSec.Waf;
+using Datadog.Trace.Logging;
+using Datadog.Trace.Telemetry;
+
+namespace Datadog.Trace.Util;
+
+/// <summary>
+/// Beware that this type is not thread safe and should be used with [ThreadStatic]
+/// </summary>
+internal class UnmanagedMemoryPoolFactory
+{
+    private static readonly IDatadogLogger Log = DatadogLogging.GetLoggerFor(typeof(UnmanagedMemoryPoolFactory));
+
+    // Statistics
+    private static int _fastPoolCount = 0;
+    private static int _slowPoolCount = 0;
+
+    public static IUnmanagedMemoryPool GetPool(int blockSize, int poolSize)
+    {
+        if (_fastPoolCount < WafConstants.MaxUnmanagedPools)
```

Comment:
I'm not a fan of this change as-is. The `UnmanagedMemoryPoolFactory` is meant to be a _utility_ class, i.e. it cam be used by any part of Datadog.Trace. This makes it WAF-specific, by tying it to WAF constants. We can change that relatively easily if you parameterise the `UnmanagedMemoryPoolFactory` for example:

```csharp
internal class UnmanagedMemoryAllocatorFactory
{
    public UnmanagedMemoryAllocatorFactory(string poolName, int? maxUnmanagedPools = null)
    // ...
}
```

The `poolName` is just so that we can distinguish different pools in the logs, the `maxUnmanagedPools` allows optional unlimited pools.

Also, I think `UnmanagedMemoryPoolFactory` and `IUnmanagedMemoryPool` etc should not have "pool" in the name, as they're _not_ necessarily pooled (as shown in the changed name above)

- `UnmanagedMemoryPoolFactory` -> `UnmanagedMemoryAllocatorFactory`
- `IUnmanagedMemoryPool` -> `IUnmanagedMemoryAllocator`
- `UnmanagedMemoryPoolSlow` -> `UnpooledUnmanagedMemoryAllocator`

The names are uglier, but they're more accurate at least 😉